### PR TITLE
* Split protocol in actual protocol and server implementation

### PR DESCRIPTION
**Pull request by mvancanneyt**
Ryan,

As I wrote some time ago, I split the lspprotocol package so it contains only the actual protocol classes. 
It no longer depends on any functionality of Lazarus itself.

There is now a lspserver package which contains the actual server implementation. 
This package has a dependency on Lazarus, obviously.

This step was necessary to be able to create a lspclient package which has no dependencies on the Lazarus IDE, allowing to write a LSP client. 

I tested the changed version in VSCode and all works well.

I have already started the lspclient package, but that will be part of a separate pull request.

### CODE REVIEW

**Comment by genericptr** [comment]
scanning over this the only thing I saw that jumped out at me was the code actions seem to be missing. In both files I don't see the original code so did it get lost?

**Comment by mvancanneyt** [comment]
There were no code actions server-side ? 
That was implemented client-side in the VSCode plugin.

**Comment by genericptr** [comment]
I didn't implement any yet but the infrastructure was there if we wanted to. The fact the code actions you did were implemented on the client seems to be something to do with VSCode but in theory they should be on the server like the rest of the functionality.

Did you end up keeping any of those code actions btw? If I ever do any code actions I don't want to duplicate them on all the clients (defeats the purpose of the LSP) so we eventually need to move them to the server do we not?

**Comment by mvancanneyt** [comment]
Ryan, 

I did not remove anything and I never implemented any specific code actions.
All I did was add a 'Code completion' code action client side, but that did not make any use of the 'CodeAction' message as it exists in the LSP protocol.

The basic message infrastructure is still there, I did not touch it. 
but the "process" method is empty since there are are no code actions.

**Comment by genericptr** [comment]
oh I'm sorry Michael I didn't look at the diff carefully enough! It's still all there. I thought I saw some code got deleted.

**Comment by mvancanneyt** [comment]
No worries. 
It is always better to check carefully :smile:

**Comment by genericptr** [comment]
Ok just had a chance to review this more and I had one request. Can you restore the unit copyright header on all the the files you split apart? I think you can just copy and paste the old one, no need to update the names/dates since no code actually changed.

**Comment by mvancanneyt** [comment]
Done.

File: src/serverprotocol/PasLS.ApplyEdit.pas [hunk 1]
```diff
@@ -1,3 +1,22 @@
+// Pascal Language Server
+// Copyright 2020 Arjan Adriaanse
```
Comment (genericptr) [inline src/serverprotocol/PasLS.ApplyEdit.pas]: Michael, I think we've added Arjan Adriaanse to files he didn't write so we should remove it. I'll make a note on the files which he didn't do, including this one.

File: src/serverprotocol/PasLS.CodeAction.pas [hunk 2]
```diff
@@ -1,3 +1,23 @@
+// Pascal Language Server
+// Copyright 2020 Arjan Adriaanse
```
Comment (genericptr) [inline src/serverprotocol/PasLS.CodeAction.pas]: Arjan Adriaanse did not author this file.

File: src/serverprotocol/PasLS.Command.CompleteCode.pas [hunk 3]
```diff
@@ -1,3 +1,22 @@
+// Pascal Language Server
+// Copyright 2020 Arjan Adriaanse
```
Comment (genericptr) [inline src/serverprotocol/PasLS.Command.CompleteCode.pas]: Arjan Adriaanse did not author this file.

File: src/serverprotocol/PasLS.Command.FormatCode.pas [hunk 4]
```diff
@@ -1,3 +1,22 @@
+// Pascal Language Server
+// Copyright 2020 Arjan Adriaanse
```
Comment (genericptr) [inline src/serverprotocol/PasLS.Command.FormatCode.pas]: Arjan Adriaanse did not author this file.

File: src/serverprotocol/PasLS.Command.InvertAssignment.pas [hunk 5]
```diff
@@ -1,3 +1,22 @@
+// Pascal Language Server
+// Copyright 2020 Arjan Adriaanse
```
Comment (genericptr) [inline src/serverprotocol/PasLS.Command.InvertAssignment.pas]: Arjan Adriaanse did not author this file.

File: src/serverprotocol/PasLS.Command.RemoveEmptyMethods.pas [hunk 6]
```diff
@@ -1,3 +1,22 @@
+// Pascal Language Server
+// Copyright 2020 Arjan Adriaanse
```
Comment (genericptr) [inline src/serverprotocol/PasLS.Command.RemoveEmptyMethods.pas]: Arjan Adriaanse did not author this file.

File: src/serverprotocol/PasLS.Diagnostics.pas [hunk 7]
```diff
@@ -1,3 +1,22 @@
+// Pascal Language Server
+// Copyright 2020 Arjan Adriaanse
```
Comment (genericptr) [inline src/serverprotocol/PasLS.Diagnostics.pas]: Arjan Adriaanse did not author this file.

File: src/serverprotocol/PasLS.DocumentSymbol.pas [hunk 8]
```diff
@@ -1,3 +1,22 @@
+// Pascal Language Server
+// Copyright 2020 Arjan Adriaanse
```
Comment (genericptr) [inline src/serverprotocol/PasLS.DocumentSymbol.pas]: Arjan Adriaanse did not author this file.

File: src/serverprotocol/PasLS.ExecuteCommand.pas [hunk 9]
```diff
@@ -1,3 +1,22 @@
+// Pascal Language Server
+// Copyright 2020 Arjan Adriaanse
```
Comment (genericptr) [inline src/serverprotocol/PasLS.ExecuteCommand.pas]: Arjan Adriaanse did not author this file.

File: src/serverprotocol/PasLS.ExecuteCommand.pas [hunk 10]
```diff
@@ -1,3 +1,22 @@
...
+// Copyright 2020 Ryan Joseph
```
Comment (genericptr) [inline src/serverprotocol/PasLS.ExecuteCommand.pas]: Year should be 2022

File: src/serverprotocol/PasLS.Formatter.pas [hunk 11]
```diff
@@ -1,3 +1,22 @@
+// Pascal Language Server
+// Copyright 2020 Arjan Adriaanse
```
Comment (genericptr) [inline src/serverprotocol/PasLS.Formatter.pas]: I think this was your file in 2023 actually?

File: src/serverprotocol/PasLS.Hover.pas [hunk 12]
```diff
@@ -1,3 +1,22 @@
+// Pascal Language Server
+// Copyright 2020 Arjan Adriaanse
```
Comment (genericptr) [inline src/serverprotocol/PasLS.Hover.pas]: Arjan Adriaanse did not author this file.

File: src/serverprotocol/PasLS.InlayHint.pas [hunk 13]
```diff
@@ -1,3 +1,22 @@
+// Pascal Language Server
+// Copyright 2020 Arjan Adriaanse
```
Comment (genericptr) [inline src/serverprotocol/PasLS.InlayHint.pas]: Arjan Adriaanse did not author this file.

File: src/serverprotocol/PasLS.InlayHint.pas [hunk 14]
```diff
@@ -1,3 +1,22 @@
...
+// Copyright 2020 Ryan Joseph
```
Comment (genericptr) [inline src/serverprotocol/PasLS.InlayHint.pas]: year should be 2022

File: src/serverprotocol/PasLS.RemoveEmptyMethods.pas [hunk 15]
```diff
@@ -1,3 +1,22 @@
+// Pascal Language Server
+// Copyright 2020 Arjan Adriaanse
```
Comment (genericptr) [inline src/serverprotocol/PasLS.RemoveEmptyMethods.pas]: I think this was your file.

File: src/serverprotocol/PasLS.SignatureHelp.pas [hunk 16]
```diff
@@ -1,3 +1,22 @@
+// Pascal Language Server
+// Copyright 2020 Arjan Adriaanse
```
Comment (genericptr) [inline src/serverprotocol/PasLS.SignatureHelp.pas]: Arjan Adriaanse did not author this file.

File: src/serverprotocol/PasLS.Workspace.pas [hunk 17]
```diff
@@ -1,3 +1,22 @@
+// Pascal Language Server
+// Copyright 2020 Arjan Adriaanse
```
Comment (genericptr) [inline src/serverprotocol/PasLS.Workspace.pas]: Arjan Adriaanse did not author this file.

**Review by genericptr — CHANGES_REQUESTED** [review]
I think you copy and pasted too much :) Sorry to bother you with this but I don't want to get copyright/credits wrong. Thanks Michael.

**Comment by mvancanneyt** [comment]
Ryan,

I corrected it.

For the future, you have my express permission to change such things yourself if you feel it is important.

Personally, I could not care less about copyright info.

**Review by genericptr — APPROVED** [review]
Good enough for now, thanks Michael. :)

**Pull request merged**